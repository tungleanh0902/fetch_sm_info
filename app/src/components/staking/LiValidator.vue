<template>
  <tr
    class="li-validator"
    :data-name="validator.name"
    @click="
      $router.push({
        name: 'validator',
        params: { validator: validator.operatorAddress },
      })
    "
  >
    <td>{{ index + 1 }}</td>
    <td class="hide-xs">
      <div class="status-container">
        <span
          :class="validator.status | toLower"
          class="validator-status"
          :title="validator.statusDetailed"
        >
          {{ validator.status }}
        </span>
      </div>
    </td>
    <td class="data-table__row__info">
      <Avatar
        v-if="!validator || !validator.picture || validator.picture === 'null'"
        class="li-validator-image"
        alt="generic validator logo - generated avatar from address"
        :address="validator.operatorAddress"
      />
      <img
        v-else-if="validator && validator.picture"
        :src="validator.picture"
        class="li-validator-image"
        :alt="`validator logo for ` + validator.name"
      />
      <div class="validator-info">
        <h3 class="li-validator-name">
          {{ validator.name }}
        </h3>
        <div v-if="delegation.amount > 0">
          <h4>
            {{ delegation.amount | bigFigureOrShortDecimals }}
          </h4>
          <h5
            v-if="
              rewards.find(
                (reward) =>
                  reward.denom === stakingDenom && reward.amount > 0.001
              )
            "
          >
            <span
              >+{{
                filterStakingDenomReward() | bigFigureOrShortDecimals
              }}</span
            >
          </h5>
        </div>
      </div>
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'expectedReturns' }">
      {{
        validator.expectedReturns
          ? bigFigureOrPercent(validator.expectedReturns)
          : `--`
      }}
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'voting-power' }">
      {{ validator.votingPower | bigFigureOrPercent }}
    </td>
  </tr>
</template>

<script>
import { bigFigureOrPercent, bigFigureOrShortDecimals } from "scripts/num"
import Avatar from "common/Avatar"

export default {
  name: `li-validator`,
  components: {
    Avatar,
  },
  filters: {
    toLower: (text) => text.toLowerCase(),
    bigFigureOrShortDecimals,
    bigFigureOrPercent,
  },
  props: {
    validator: {
      type: Object,
      required: true,
    },
    /* istanbul ignore next */
    delegation: {
      type: Object,
      default: () => ({}),
    },
    /* istanbul ignore next */
    rewards: {
      type: Array,
      default: () => ({}),
    },
    index: {
      type: Number,
      required: true,
    },
    /* istanbul ignore next */
    showOnMobile: {
      type: String,
      default: () => "returns",
    },
    stakingDenom: {
      type: String,
      default: "",
    },
  },
  methods: {
    bigFigureOrPercent,
    bigFigureOrShortDecimals,
    filterStakingDenomReward() {
      const stakingDenomRewards = this.rewards.filter(
        (reward) => reward.denom === this.stakingDenom
      )
      return stakingDenomRewards.length > 0 ? stakingDenomRewards[0].amount : 0
    },
  },
}
</script>
<style scoped>
.li-validator {
  padding: 0.5rem 1rem;
  margin-bottom: 0.25rem;
  border-bottom: 1px solid var(--bc-dim);
  border-radius: 0.25rem;
}

.validator-info {
  display: flex;
  flex-direction: column;
  padding-left: 1rem;
  text-overflow: ellipsis;
}

.li-validator h4,
.li-validator h5 {
  font-size: var(--sm);
  display: inline-block;
}

.li-validator h5 {
  padding-left: 0.5rem;
  color: var(--success);
}

.li-validator:hover {
  cursor: pointer;
  background: var(--app-fg-hover);
  color: var(--bright);
}

.li-validator-name {
  font-size: 1rem;
  line-height: 18px;
  font-weight: 500;
  color: var(--bright);
  display: inline-block;
  max-width: 20rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.li-validator-image {
  border-radius: 0.25rem;
  height: 2.5rem;
  width: 2.5rem;
  min-height: 2.5rem;
  min-width: 2.5rem;
  border: 1px solid var(--bc-dim);
}

.validator-status {
  text-transform: uppercase;
  font-size: 10px;
  font-weight: 600;
  border: 2px solid;
  padding: 2px 4px;
  border-radius: 0.25rem;
}

.validator-status.inactive {
  color: var(--warning);
  border-color: var(--warning);
}

.validator-status.active {
  color: var(--success);
  border-color: var(--success);
}

@media screen and (max-width: 768px) {
  .li-validator-name {
    max-width: 11rem;
  }
}
</style>
