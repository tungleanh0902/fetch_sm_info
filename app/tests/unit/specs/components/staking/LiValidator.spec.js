import { shallowMount, createLocalVue } from "@vue/test-utils"
import VueApollo from "vue-apollo"
import LiValidator from "src/components/staking/LiValidator"

const localVue = createLocalVue()
localVue.directive(`tooltip`, () => {})
localVue.use(VueApollo)

describe(`LiValidator`, () => {
  let wrapper, $store

  const validator = {
    networkId: "cosmoshub",
    operatorAddress: "cosmosvaloper1abcdefghijklmnop",
    consensusPubkey: "cosmosvalconspub11234567890",
    jailed: false,
    details: "The Validator",
    website: "http://validator.com/",
    identity: "ABCDEFGHIJKL",
    votingPower: "0.014",
    startHeight: 190459,
    uptimePercentage: "1",
    tokens: "123456789",
    updateTime: "2019-05-31T23:54:12.176123478Z",
    commission: "0.070000000000000000",
    maxCommission: "0.150000000000000000",
    maxChangeCommission: "0.030000000000000000",
    commissionLastUpdate: null,
    height: 12345,
    status: "ACTIVE",
    statusDetailed: "active",
    delegations: null,
    selfStake: {
      amount: 123,
    },
    expectedReturns: "123",
    customized: null,
    tombstoned: null,
    keybaseId: null,
    lastUpdated: null,
    minSelfDelegation: null,
    profileUrl: null,
    userName: null,
    picture: "picture.jpg",
    name: "",
    userShares: {
      amount: 123,
    },
  }

  const index = 1

  beforeEach(() => {
    $store = {
      state: {
        pool: {
          pool: {
            bonded_tokens: 1000,
          },
        },
      },
    }

    wrapper = shallowMount(LiValidator, {
      localVue,
      propsData: {
        validator,
        rewards: [
          {
            amount: 1,
            denom: `token1`,
          },
        ],
        index,
        showOnMobile: "returns",
      },
      mocks: {
        $store,
      },
      stubs: [`router-link`],
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should show the voting power`, () => {
    expect(wrapper.html()).toContain(`1.40%`)
  })

  it(`should gravatar when no avatar`, () => {
    wrapper.setProps({
      validator: {
        ...validator,
        picture: "",
      },
    })
    expect(wrapper.find("avatar-stub").exists())
  })

  it(`should filter the staking denom reward`, () => {
    const rewards = [
      {
        amount: 1,
        denom: `TOKEN1`,
      },
      {
        amount: 2,
        denom: `TOKEN2`,
      },
      {
        amount: 3,
        denom: `TOKEN3`,
      },
    ]
    const self = {
      stakingDenom: `TOKEN1`,
      rewards: rewards,
    }
    const stakingDenomReward = LiValidator.methods.filterStakingDenomReward.call(
      self
    )
    expect(stakingDenomReward).toEqual(1)
  })
})
