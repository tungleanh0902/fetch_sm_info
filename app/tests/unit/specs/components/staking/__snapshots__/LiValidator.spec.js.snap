// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiValidator has the expected html structure 1`] = `
<tr
  class="li-validator"
  data-name=""
>
  <td>
    2
  </td>
   
  <td
    class="hide-xs"
  >
    <div
      class="status-container"
    >
      <span
        class="validator-status active"
        title="active"
      >
        
        ACTIVE
      
      </span>
    </div>
  </td>
   
  <td
    class="data-table__row__info"
  >
    <img
      alt="validator logo for "
      class="li-validator-image"
      src="picture.jpg"
    />
     
    <div
      class="validator-info"
    >
      <h3
        class="li-validator-name"
      >
        
        
      
      </h3>
       
      <!---->
    </div>
  </td>
   
  <td
    class="hide-xs"
  >
    
    12,300.00%
  
  </td>
   
  <td
    class="hide-xs"
  >
    
    1.40%
  
  </td>
</tr>
`;
